src/cript/api/utils/save_helper.py

```diff
@@ -1,50 +1,29 @@
-from cript.api.exceptions import CRIPTAPISaveError
-
-
-def brute_force_save(project, bad_uuid: str) -> None:
+def fix_node_save(api, node, response, known_uuid):
     """
-    the API class tries to save the project, if the API gives a `Bad UUID` error
-    then it will try to brute_force_save it
+    Helper function, that attempts to fix a bad node.
+    And if it is fixable, we resave the entire node.
 
-    try to save project
-        if `Bad UUID error`
-            find the node that has to be saved first before being referred to as just UUID in JSON
-                save that material
-                strip that node and replace them with the saved UUID only throughout the whole JSON
+    Returns set of known uuids, if fixable, otherwise False.
     """
-    from cript.api.api import _get_global_cached_api
-
-    while True:
-        try:
-            # find the node that has bad UUID error
-            node_to_save_first = find_node_by_uuid(project_node=project, uuid=bad_uuid)
-
-            # save the node that was unsaved to the API
-            api = _get_global_cached_api()
-            api.send_post_request(node=node_to_save_first)
-
-            # try to save the whole project again and see if it'll work this time
-            api.save(project=project)
-
-            # if everything went successfully and everything saved well, then exist
-            break
-
-        except CRIPTAPISaveError as error:
-            # if there is a bad UUID error then it will catch it here
-            # get the bad UUID, set the bad UUID, and try brute force saving again
-
-            error_message = str(error)
-            if error_message.startswith("API responded with 'http:400 Bad uuid: "):
-                bad_uuid = get_bad_uuid_from_error_message(error_message=error_message)
-
-            # there was an error with the save, so loop over again and try again
-
-            # if we are not handling the error then raise it
-            else:
-                raise error
-
-
-def get_bad_uuid_from_error_message(error_message: str) -> str:
+    assert response["code"] in (400, 409)
+    if response["error"].startswith("Bad uuid:") or response["error"].strip().startswith("Duplicate uuid:"):
+        missing_uuid = get_uuid_from_error_message(response["error"])
+        missing_node = find_node_by_uuid(node, missing_uuid)
+
+        # Now we save the bad node extra.
+        # So it will be known when we attempt to save the graph again.
+        # Since we pre-saved this node, we want it to be UUID edge only the next JSON.
+        # So we add it to the list of known nodes
+        known_uuid.union(api._internal_save(missing_node, known_uuid))  # type: ignore
+        # The missing node, is now known to the API
+        known_uuid.add(missing_uuid)
```

Comment:
I think `known_uuid` might be better named `saved_nodes_uuid`, but a name doesn't make much of a difference